pyro/distributions/dequantized.py

```diff
@@ -0,0 +1,93 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+
+import torch
+from torch.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class DequantizedDistribution(TorchDistribution):
+    """
+    Dequantized distribution over relaxed counts.
+
+    The density of this distribution linearly interpolates the density of
+    ``base_dist``, except in the interval ``[0,1]`` where the weight of 0 is
+    doubled to account for asymmetry.
+
+    Given a distribution ``base_dist`` over nonnegative integers, this creates
+    a distribution over positive reals such that randomly rounded samples from
+    this distribution will be distributed approximately according to
+    ``base_dist``.
+
+    For example the following two models result in approximately identical
+    distributions over ``z``, conditioned on inputs ``n, p``.
+
+        # Model 1.
+        z ~ Binomial(n, p)
+
+        # Model 2.
+        r ~ DequantizedDistribution(Binomial(n, p))
+        b ~ Bernoulli(1 + floor(r) - r)      # Quantize.
+        z = floor(r) + b
+
+    Earth mover distance error is upper bounded by 1/2.
+
+    :param base_dist: A distribution with
+        ``.support == constraints.nonnegative_integer`` and
+        ``.event_shape == ()``.
+    :type base_dist: ~pyro.distributions.TorchDistribution
+    """
+    arg_constraints = {}
+    support = constraints.positive  # Note lack of upper bound.
+    has_rsample = True
+
+    def __init__(self, base_dist, validate_args=None):
+        self.base_dist = base_dist
+        super().__init__(base_dist.batch_shape, validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(DequantizedDistribution, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new.base_dist = self.base_dist.expand(batch_shape)
+        super(DequantizedDistribution, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @torch.no_grad()
+    def sample(self, sample_shape=torch.Size()):
+        value = self.base_dist.sample(sample_shape)
+        # Add uniform triangular noise in [-1,1]
```

Comment:
Looking like this will not make Model1 and Model2 equivalent:
p_model2 (z) = E_{u in (0, 1)} [u*p(z) + (1-u)*p(z-1)] + E_{u in (0, 1)} [u*p(z) + (1-u)*p(z+1)] = p(z) + (p(z-1) + p(z+1))/2 != p(z)

Reply:
Agreed, I think the best we can hope for is approximate equivalence with a good bound of 0.5 on earth mover distance error. It might be possible to implement distribution-specific dequantized distributions with tighter bounds, but I think it will be difficult to make their gradients well-behaved at the boundaries (0 and total_count).

Reply:
I think we can come up with an equivalent procedure by uniformly generating the noise in the interval (-C / p(n), C / p(n)) where C is small enough such that the interval belongs to (-1, 1) for all `n`. I'll make a notebook to show the equivalent after the dinner.

Reply:
Oh good point, `torch.uniform(0,1) + torch.uniform(-1,0)` is triangular :smile:

Reply:
My point is as long as we use a noise with the density mean proportional to 1 / p(n), then quantize(dequantize(x)) has the same distribution as x. The uniform noise as in my last comment preserves that property but dequantize(x) is discontinuous. We can use something like trapezoidal but the density is not monotone (it bends up and down). The triangular noise is nice but it does not preserve that relationship. I am a bit worried about this inconsistency
```python
d = dist.DequantizedDistribution(dist.Poisson(0.1))
x = torch.linspace(-1, 5, 1000)
z = d.sample((1000000,))
z = (z - z.floor()).bernoulli() + z.floor()
plt.figure(figsize=(8, 3), dpi=300).patch.set_color('white')
plt.hist(z, bins=200, range=(-1,5), alpha=0.5)
plt.hist(d.base_dist.sample((1000000,)), bins=200, range=(-1,5), alpha=0.5);
```

Reply:
On second thought, I think it can be fine to use `triangular` here. Could you add a note so we can revisit in the future? for example,
```
# we use triangular noise here instead of other candidates because
#   + the density will be continuous
#   + the density will have similar monotonicity as base_dist
#   + the density is easy to compute
```